File: src/main/java/guru/xgm/image/blp/BlpWriterSpi.java
package guru.xgm.image.blp;

import static guru.xgm.image.blp.BlpImageSpiCommon.*;

import java.io.File;
import java.nio.file.Path;
import java.util.Locale;

import javax.imageio.ImageTypeSpecifier;
import javax.imageio.ImageWriter;
import javax.imageio.spi.ImageWriterSpi;
import javax.imageio.stream.ImageOutputStream;

/**
 * Service provider for BLP image file ImageWriter.
 *
 * @author Imperial Good
 */
public class BlpWriterSpi extends ImageWriterSpi {
    static final String WRITER_CLASS = "guru.xgm.image.blp.BlpWriter";
    static final Class<?>[] OUTPUT_TYPES = {ImageOutputStream.class,
            File.class, Path.class};
    static final String[] READER_SPI_CLASSES = {"guru.xgm.image.blp.BlpReaderSpi"};

    public BlpWriterSpi() {
        super(VENDOR, VERSION, FORMAT_NAMES, FORMAT_SUFFIXES, FORMAT_MIMES,
                WRITER_CLASS, OUTPUT_TYPES, READER_SPI_CLASSES,
                STANDARD_STREAM_METADATA_SUPPORT, NATIVE_STREAM_METADATA_NAME,
                NATIVE_STREAM_METADATA_CLASS, EXTRA_STREAM_METADATA_NAME,
                EXTRA_STREAM_METADATA_CLASS, STANDARD_IMAGE_METADATA_SUPPORT,
                NATIVE_IMAGE_METADATA_NAME, NATIVE_IMAGE_METADATA_CLASS,
                EXTRA_IMAGE_METADATA_NAME, EXTRA_IMAGE_METADATA_CLASS);
    }

    @Override
    public boolean canEncodeImage(ImageTypeSpecifier type) {
        // not at all strict for maximum usability
        return true;
    }

    @Override
    public ImageWriter createWriterInstance(Object extension) {
        return new BlpWriter(this);
    }

    @Override
    public String getDescription(Locale locale) {
        return "BLP file image writer.";
    }

}

File: src/main/kotlin/raft/war/image/blp/BlpWriterSpi.kt
package raft.war.image.blp

import java.io.File
import java.nio.file.Path
import java.util.Locale
import javax.imageio.ImageTypeSpecifier
import javax.imageio.ImageWriter
import javax.imageio.spi.ImageWriterSpi
import javax.imageio.stream.ImageOutputStream

class BlpWriterSpi : ImageWriterSpi(
    BlpImageSpiCommon.VENDOR,
    BlpImageSpiCommon.VERSION,
    BlpImageSpiCommon.FORMAT_NAMES,
    BlpImageSpiCommon.FORMAT_SUFFIXES,
    BlpImageSpiCommon.FORMAT_MIMES,
    BlpWriter::class.java.getCanonicalName(),
    arrayOf<Class<*>>(ImageOutputStream::class.java, File::class.java, Path::class.java),
    arrayOf<String>(BlpReaderSpi::class.java.getCanonicalName()),
    BlpImageSpiCommon.STANDARD_STREAM_METADATA_SUPPORT,
    BlpImageSpiCommon.NATIVE_STREAM_METADATA_NAME,
    BlpImageSpiCommon.NATIVE_STREAM_METADATA_CLASS,
    BlpImageSpiCommon.EXTRA_STREAM_METADATA_NAME,
    BlpImageSpiCommon.EXTRA_STREAM_METADATA_CLASS,
    BlpImageSpiCommon.STANDARD_IMAGE_METADATA_SUPPORT,
    BlpImageSpiCommon.NATIVE_IMAGE_METADATA_NAME,
    BlpImageSpiCommon.NATIVE_IMAGE_METADATA_CLASS,
    BlpImageSpiCommon.EXTRA_IMAGE_METADATA_NAME,
    BlpImageSpiCommon.EXTRA_IMAGE_METADATA_CLASS
) {
    override fun canEncodeImage(type: ImageTypeSpecifier?): Boolean = true
    override fun createWriterInstance(extension: Any?): ImageWriter = BlpWriter(this)
    override fun getDescription(locale: Locale?): String = "BLP file image writer."
}
